### ISSUE TITLE
Permit passing a languageId as an option to openTextDocument

### ISSUE THREAD

**Issue by pvh** [Microsoft/vscode#62749]
I'm working on a filesystem provider and I'm annoyed that all my files show up as plaintext if I use the vscode.open command.  More specifically, I have JSON data that I'm rendering as strings for the user to edit (which I later reparse into JSON and extract changes from).

I'd like a mechanism whereby I can tell VSCode in my FileSystemProvider what the languageId of that file is. Thinking about how to accomplish this without changing the current APIs too much, perhaps there could be a readFileLanguage() call that the languages services could check for when they try to decide which language to render the file as?

**Comment by jrieken** [comment]
We have `setTextDocumentLanguage` for this: https://github.com/Microsoft/vscode/blob/f5b5630d900932c1e2992f01e98d225a4b189c6e/src/vs/vscode.d.ts#L7507-L7519

We have to have things decoupled and I'd prefer to keep those things separate. I think using the function above is the easiest, otherwise you can try the declarative `languages`-section in `package.json`. It allows to bind to existing languages (like JSON) using filename patterns (like foo1, foo2, if that applies) or the first line content of the file

**Comment by pvh** [comment]
Yeah, I understand the desire to keep those separate but I can't see how I can ensure that every "vscode.open" call will get my language preferences applied. This means, for example, that if you ctrl-click a link to open a document referred to in another document it doesn't get source highlighting because there's no way (I have thought of, at least) for me to "hook" the document open call and run setTextDocumentLanguage.

Right now my work-around is that I have a custom treeview where clicks trigger a special "myextension.open", but there are quite a few ways of opening a document that don't follow that path and so don't have the right highlighting.

**Comment by jrieken** [comment]
Why are you using `vscode.open` to begin with? There is `openTextDocument` and `showTextDocument` which are the better choice. Anyways, use the `onDidOpenTextDocument`-event and check if it's one for your document that hasn't the right language yet, then change it

**Comment by pvh** [comment]
I'm not using vscode.open -- it seems the system uses that when I reopen the editor or Ctrl-Click a link to a document, but that hook is exactly what I overlooked. Thanks for the idea, it works great!
